### PR TITLE
Expanded Profile Launch and Kill

### PR DESCRIPTION
**Pull request by BlueFinBima**
Expanded the processing of Launch, Kill Process Image Name, and triggers for the profile start, stop and reset.

### CODE REVIEW

**Comment by derammo** [comment]
you want notes on this or just want me to commit it?   

at least one confirmation I need:  The copyright messages in the aircraft image modules assembly info seem weird.  Are all those images actually from you?  I thought the A-10C images were original Helios, but I wasn't there in the early days so I don't know.

File: Helios/Helios.csproj [hunk 1]
```diff
@@ -18,6 +18,9 @@
   <PropertyGroup>
     <ApplicationIcon>Helios.ico</ApplicationIcon>
   </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Development|x64'">
+    <LangVersion>latest</LangVersion>
```
Comment (derammo) [inline Helios/Helios.csproj:22]: why is this here?

File: Helios/Windows/Controls/About.xaml.cs [hunk 2]
```diff
@@ -27,7 +27,7 @@ namespace GadrocsWorkshop.Helios.Windows.Controls
     public partial class About : Window
     {
         // these are in order of being primary on the project
-        public static readonly string[] Authors = { "Gadroc", "BlueFinBima", "derammo" };
+        public static readonly string[] Authors = { "Gadroc", "derammo", "BlueFinBima" };
```
Comment (derammo) [inline Helios/Windows/Controls/About.xaml.cs:30]: I put the authors in historical order, to denote the life cycle phases of Helios :)

Comment (derammo) [reply Helios/Windows/Controls/About.xaml.cs:30]: appreciate the nod though!

Comment (BlueFinBima) [reply Helios/Windows/Controls/About.xaml.cs:30]: My preference is in contribution order - hence the acknowledgement to all that you have done.

File: Helios/Helios.csproj (hunk 1 above)
Comment (BlueFinBima) [reply Helios/Helios.csproj:22]: The use of default value expressions caused my VS to barf because of the language level.  Happy for you to drop this, but I'll still need  to put this into my branch until I can upgrade.

**Comment by BlueFinBima** [comment]
> at least one confirmation I need: The copyright messages in the aircraft image modules assembly info seem weird. Are all those images actually from you? I thought the A-10C images were original Helios, but I wasn't there in the early days so I don't know.

TBH, I do not know the heritage of much of the imagery.  It is obviously possible that some of it came from Gadroc, or possibly Loz (whoever that is/was), but it might also have ownership claimed by ED.  That is why I wimped out and didn't indicate who had copyright, but I wanted to indicate that there probably was one or more copyright holders.
I guess an alternate is to state that no copyright is claimed by the project, but the information may have copyright claimed / asserted.

**Comment by derammo** [comment]
I think it says BlueFinBima Project as the copyright holder or is that a mistake in the web UI

File: Helios/Helios.csproj (hunk 1 above)
Comment (derammo) [reply Helios/Helios.csproj:22]: So you are saying you are on older Visual Studio or something and it is selecting a different language version?  That seems super weird since selecting 4.7.2 .NET should force language version 7?

**Comment by BlueFinBima** [comment]
> I think it says BlueFinBima Project as the copyright holder or is that a mistake in the web UI Sent from my iPhone

In 1.4, all it said in the Assembly Information was:
image

Regardless of what it does say today, the intent should be to indicate that it is probably copyright, but that the BlueFinBima project is not claiming to be the copyright holder.  Prior to me pulling  it into it's own DLL, it was in Helios.dll, and Gadroc asserted copyright on that.

What do you propose / suggest we do?

File: Helios/Helios.csproj (hunk 1 above)
Comment (BlueFinBima) [reply Helios/Helios.csproj:22]: Unfortunately, target-typed default literals were not in 7.0, they came in at 7.1, so I have to set "latest minor version (latest)" to coax VS to use that language level.  Seems to be independent of the ,Net level.  I think you assert the need to build on 2019, but there are a couple of extensions for uControllers which do not work on 2019 so I am hanging on to 2017 until I can find an alternative solution.
Language level 7.**1** is the current minimum for Helios 1.6.

**Comment by derammo** [comment]
I just thought it was a typo.   I did not realize you were a project.  I mean, personally yes, and I am in favor of self-improvement.  But I thought it was a typo in this context.  We should put "Helios Contributors" to be consistent with all the other places where I added this text to indicate people other than Gadroc have contributed and you can't take this code private without their consent.  That is the only reason I care about Copyright.   I want to prevent someone getting permission from Gadroc to make this a commercial product and re-release under a non-GPL license because a bunch of our code incorrectly asserted it is Copyright Gadroc.

**Comment by derammo** [comment]
You didn't say if you want notes on the code in this review.  Is that a no? :)

**Comment by BlueFinBima** [comment]
If you have the time, then yes to the notes.

File: Helios/Interfaces/Profile/ProfileInterface.cs [hunk 3]
```diff
@@ -243,21 +372,54 @@ private void CheckResetMonitors(IList<StatusReportItem> newReport)
...
         /// <param name="newReport"></param>
         private void CheckExecutablePaths(IList<StatusReportItem> newReport)
         {
-            // TODO if any launch or kill actions are configured, check paths and access
-            // TODO if incorrect, report warning or error (to stop launch) in StatusReportItem
-            // TODO if correct, report detail also in StatusReportItem
-            _ = newReport;
+            foreach(HeliosBinding _binding in OutputBindings)
+            {
+                if(_binding.Action.ActionID == "launch application")
```
Comment (derammo) [inline Helios/Interfaces/Profile/ProfileInterface.cs:389]: switch on binding.Action.ActionID and log in default case that you found an unimplemented / unexpected command

C# does switch/case on strings

Comment (derammo) [reply Helios/Interfaces/Profile/ProfileInterface.cs:389]: btw probably fine to intentionally throw Null exception (i.e. do nothing) if binding.Action is null.  I think that would be a structural flaw that SHOULD come back as an exception stack trace from the field.   So I would leave it unguarded like you have it.

**Comment by derammo** [comment]
I noted things I saw.  Came out a bit more rude than intended, but just because of expedience and being tired.  Sorry.

File: Aircraft A-10C/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -10,7 +10,7 @@
 [assembly: AssemblyConfiguration("")]
 [assembly: AssemblyCompany("github.com/HeliosVirtualCockpit")]
 [assembly: AssemblyProduct("Helios")]
-[assembly: AssemblyCopyright("Copyright (c) 2019 BlueFinBima")]
+[assembly: AssemblyCopyright("Copyright (c) 2019 BlueFinBima Project")]
```
[the remark this answers is not in the document]
Comment (BlueFinBima) [reply Aircraft A-10C/Properties/AssemblyInfo.cs:13]: "Helios Contributors" it is...

**Comment by BlueFinBima** [comment]
> I noted things I saw. Came out a bit more rude than intended, but just because of expedience and being tired. Sorry.

Not a problem, if I get all sensitive about this, you just won't get a Christmas Card from me.

File: Helios/Interfaces/Profile/ProfileInterface.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (BlueFinBima) [reply Helios/Interfaces/Profile/ProfileInterface.cs:389]: I'd already changed this to 7.1, so moved into Platforms.properties and PlatformsForPlugin.properties

[the remark this answers is not in the document]
Comment (BlueFinBima) [reply Helios/Interfaces/Profile/ProfileInterface.cs:389]: > everyone reading it will think it refers to a field. Just use "lowerCamelCase".
Not everyone... programming troglodytes like myself wouldn't infer much from this.  Incorrect Underscores obliterated.

File: Helios/Interfaces/Profile/ProfileInterface.cs [hunk 5]
```diff
@@ -278,5 +440,119 @@ public IEnumerable<StatusReportItem> PerformReadyCheck()
             CheckExecutablePaths(items);
             return items;
         }
+
+        private bool ValidateExecPath(string _proposedPath){
```
[the remark this answers is not in the document]
Comment (BlueFinBima) [reply Helios/Interfaces/Profile/ProfileInterface.cs:444]: Purged

File: Helios/Interfaces/Profile/ProfileInterface.cs [hunk 6]
```diff
@@ -278,5 +440,119 @@ public IEnumerable<StatusReportItem> PerformReadyCheck()
...
+            // validating the first file or URI
+            //
+
+            // Resolve environment variables
+            _proposedPath = Environment.ExpandEnvironmentVariables(_proposedPath);
+
+            RegexOptions options = RegexOptions.IgnoreCase | RegexOptions.IgnorePatternWhitespace;
+            MatchCollection _matches;
+
+            _matches = Regex.Matches(_proposedPath, "(?:\\\")([^\"]*)(?:\\\")",options);  // Extract anything which is enclosed in escaped double quotes
```
[the remark this answers is not in the document]
Comment (BlueFinBima) [reply Helios/Interfaces/Profile/ProfileInterface.cs:464]: I'm hate to cause .net virtual machine to be brought to it's knees due to my code, so changed the Regex implementation.

File: Helios/Interfaces/Profile/ProfileInterface.cs [hunk 7]
```diff
@@ -278,5 +440,119 @@ public IEnumerable<StatusReportItem> PerformReadyCheck()
...
+                // nothing enclosed in escaped double quotes
+                //
+                // parse assuming not a URI
+                try
+                {
+                    _proposedPath = Path.GetFullPath(_proposedPath);
+                }
+                catch (ArgumentException Ex)
+                {
+                    Logger.Debug("GetFullPath Parsing for \"" + _proposedPath + "\" threw exception \"" + Ex.ToString() + "\"");
```
[the remark this answers is not in the document]
Comment (BlueFinBima) [reply Helios/Interfaces/Profile/ProfileInterface.cs:477]: b)  wan't aware of the "+" having this affect.  
Happy to comply with the way the logger expects.  I should have checked implementation before using it.

File: Helios/Interfaces/Profile/ProfileInterface.cs (hunk 3 above)
Comment (BlueFinBima) [reply Helios/Interfaces/Profile/ProfileInterface.cs:389]: I think if we wish to validate all of the actions, then this code needs to go elsewhere.  Since CheckExecutablePaths has a very particular purpose, I propose leaving this code as is.

**Comment by BlueFinBima** [comment]
@derammo,  I really appreciate you taking the time to beat some sense into me.  I really do appreciate it.  The last problem I have (for the time being at least) is how I add an additional commit into this pull request because I do not want to lose these comments and the link to the code changes you suggested.

**Comment by BlueFinBima** [comment]
Three more commits added to address review comments (hopefully).

File: Helios/Interfaces/Profile/ProfileInterface.cs (hunk 3 above)
Comment (derammo) [reply Helios/Interfaces/Profile/ProfileInterface.cs]: I just mean if then else if is inferior to switch / case for readability and has less chance of being broken by a careless edit later.   if these are not all the actions, then of course don't log in the default :)

but if then else if on the same variable is still a switch/case in my book.

**Comment by derammo** [comment]
> @derammo,  I really appreciate you taking the time to beat some sense into me.  I really do appreciate it.  The last problem I have (for the time being at least) is how I add an additional commit into this pull request because I do not want to lose these comments and the link to the code changes you suggested.

please just keep committing to this branch until it is done it will update the PR automatically I think? 

do please stop merging stuff from master though :).   This should be one way traffic to reduce the risk level.

File: Helios/Interfaces/Profile/ProfileInterface.cs [hunk 8]
```diff
@@ -133,13 +154,122 @@ protected override void DetachFromProfileOnMainThread(HeliosProfile oldProfile)
...
+                                expandedArgs += " " + Environment.ExpandEnvironmentVariables(matchItem.ToString());
+                            }
+                            else
+                            {
+                                expandedArgs += Environment.ExpandEnvironmentVariables(e.Value.StringValue.Substring(matchCursor, matchItem.Index - matchCursor));
+                                matchCursor = matchItem.Index + matchItem.Length;
+                                expandedArgs += " " + Environment.ExpandEnvironmentVariables(matchItem.ToString());
+                            }
+                        }
+                        Logger.Debug("Double Quoted item found {ItemText}", matchItem.ToString());
```
Comment (derammo) [inline Helios/Interfaces/Profile/ProfileInterface.cs:231]: just "matchItem"   as ToString() is called by the logger

File: Helios/Interfaces/Profile/ProfileInterface.cs [hunk 9]
```diff
@@ -133,13 +154,122 @@ protected override void DetachFromProfileOnMainThread(HeliosProfile oldProfile)
...
+            {
+                Logger.Error(ex,"Error caught launching external application (path={ExpandedPath})", expandedPath);
+            }
+        }
+
+        void KillApplication_Execute(object action, HeliosActionEventArgs e)
+        {
+            try
+            {
+                Process[] _localProcessesByName = Process.GetProcessesByName(e.Value.StringValue);
```
Comment (derammo) [inline Helios/Interfaces/Profile/ProfileInterface.cs:263]: variable name

File: Helios/Interfaces/Profile/ProfileInterface.cs [hunk 10]
```diff
@@ -37,6 +39,14 @@ public class ProfileInterface : HeliosInterface, IStatusReportNotify, IResetMoni
         // our implementation of IStatusReportNotify
         private readonly StatusReportNotifyAsyncOnce _statusReportNotify;
 
+        // definition for profile related triggers
+        private HeliosTrigger _profileStartedTrigger;
+        private HeliosTrigger _profileResetTrigger;
+        private HeliosTrigger _profileStoppedTrigger;
+
+        private static Regex _rxDoubleQuotes = new Regex("(?:\\\")([^\"]*)(?:\\\")", RegexOptions.IgnoreCase | RegexOptions.IgnorePatternWhitespace);
+        private static Regex _rxExe = new Regex("(" + Environment.GetEnvironmentVariable("PATHEXT").Replace(";", "|").Replace(".", "\\.") + ")", RegexOptions.IgnoreCase | RegexOptions.IgnorePatternWhitespace);
```
Comment (derammo) [inline Helios/Interfaces/Profile/ProfileInterface.cs:48]: GetEnvironmentVariable can return null.   so you must include something like 

GetEnvironmentVariable("PATHEXT")?.Replace(bla bla bla) ?? "" 

to gracefully handle Null, particularly since this is executed at an unknown time (static init whenever) which is a super nasty thing to debug

File: Helios/Interfaces/Profile/ProfileInterface.cs (hunk 3 above)
Comment (BlueFinBima) [reply Helios/Interfaces/Profile/ProfileInterface.cs]: All for readability

File: Helios/Interfaces/Profile/ProfileInterface.cs (hunk 8 above)
Comment (BlueFinBima) [reply Helios/Interfaces/Profile/ProfileInterface.cs:231]: removed

File: Helios/Interfaces/Profile/ProfileInterface.cs (hunk 9 above)
Comment (BlueFinBima) [reply Helios/Interfaces/Profile/ProfileInterface.cs:263]: Whoops

File: Helios/Interfaces/Profile/ProfileInterface.cs (hunk 10 above)
Comment (BlueFinBima) [reply Helios/Interfaces/Profile/ProfileInterface.cs:48]: null handled.

File: Helios/Interfaces/Profile/ProfileInterface.cs [hunk 11]
```diff
@@ -278,5 +440,119 @@ public IEnumerable<StatusReportItem> PerformReadyCheck()
...
+                {
+                    Logger.Debug("GetFullPath Parsing for \"" + _matches[0].Groups[1].ToString() + "\" threw exception \"" + Ex.Message + "\"");
+                    return false;
+                }
+                catch(NotSupportedException Ex)
+                {
+                    // If we have a URI and attempt to get the full path, then this is probably what caused us to be here
+                    Logger.Debug("GetFullPath Parsing for \"" + _matches[0].Groups[1].ToString() + "\" threw exception \"" + Ex.Message + "\"");
+                }
+                try
```
[the remark this answers is not in the document]
Comment (BlueFinBima) [reply Helios/Interfaces/Profile/ProfileInterface.cs]: Done.  Not used extract method before. Nice and easy

**Comment by derammo** [comment]
not actually done via this merge, not sure why this was marked as fully merged, but I did not intend to mark this closed.  still work in progress

**Comment by derammo** [comment]
this code was split up into two things:
https://github.com/HeliosVirtualCockpit/Helios/commit/2f5f97506303c5a3a6177dcbdf499083a2908b44

and a new plugin to hold the process control (start, kill) feature, which is still in progress.  I will update when I have a commit for that.

**Comment by derammo** [comment]
ok the plugin is finished here
2a04fb88dd3bff750fe585fa3a29cc378b4c03cc
and installed here
474bd9eef61ee6c0a5b1e13c882934ea4b9c5d6e
